doc/internal/man3/OSSL_EVENT.pod

```diff
@@ -0,0 +1,217 @@
+=pod
+
+=head1 NAME
+
+OSSL_EVENT_QUEUE, OSSL_EVENT, OSSL_EVENT_SUBSCRIPTION,
+ossl_event_set, ossl_event_new, ossl_event_free,
+ossl_event_get_type, ossl_event_get_priority, ossl_event_get_when,
+ossl_event_get0_payload,
+ossl_event_queue_new, ossl_event_queue_free,
+ossl_event_queue_schedule, ossl_event_queue_delete,
+ossl_event_queue_time_until, ossl_event_queue_time_until_next,
+ossl_event_queue_postpone, ossl_event_queue_postpone_until,
+ossl_event_queue_dispatch,
+ossl_event_queue_subscribe, ossl_event_queue_unsubscribe
+- event and timer queue
+
+=head1 SYNOPSIS
+
+ #include "internal/event_queue.h"
+
+ typedef OSSL_EVENT;
+ typedef OSSL_EVENT_QUEUE;
+ typedef OSSL_EVENT_SUBSCRIPTION;
+
+ OSSL_EVENT_QUEUE *ossl_event_queue_new(void);
+ void ossl_event_queue_free(OSSL_EVENT_QUEUE *queue);
+
+ void ossl_event_set(OSSL_EVENT *event, uint32_t type, uint32_t priority,
+                     OSSL_TIME when, void *ctx,
+                     void *payload, size_t payload_size);
+ OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                            OSSL_TIME when, void *ctx,
+                            void *payload, size_t payload_size);
+ void ossl_event_free(OSSL_EVENT *event);
+
+ uint32_t ossl_event_get_type(const OSSL_EVENT *event);
+ uint32_t ossl_event_get_priority(const OSSL_EVENT *event);
+ OSSL_TIME ossl_event_get_when(const OSSL_EVENT *event);
+ void *ossl_event_get0_payload(const OSSL_EVENT *event, size_t *length);
+
+ int ossl_event_queue_schedule(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+ int ossl_event_queue_delete(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event);
+
+ OSSL_TIME ossl_event_queue_time_until(OSSL_EVENT *event);
+ OSSL_TIME ossl_event_queue_time_until_next(const OSSL_EVENT_QUEUE *queue);
+
+ int ossl_event_queue_postpone(OSSL_EVENT_QUEUE *queue, OSSL_EVENT *event,
+                               OSSL_TIME how_long);
+ int ossl_event_queue_postpone_until(OSSL_EVENT_QUEUE *queue,
+                                     OSSL_EVENT *event,
+                                     OSSL_TIME when);
+
+ int ossl_event_queue_dispatch(OSSL_EVENT_QUEUE *queue);
+
+ OSSL_EVENT_SUBSCRIPTION *ossl_event_queue_subscribe(OSSL_EVENT_QUEUE *queue,
+                                                     uint32_t type,
+                                                     ossl_event_callback_fn *cb,
+                                                     void *cbarg);
+ void ossl_event_queue_unsubscribe(OSSL_EVENT_SUBSCRIPTION *sub);
+
+=head1 DESCRIPTION
+
+These functions implement an event queue.
+
+=head2 Event
+
+An event is a small structure, B<OSSL_EVENT>, carrying information:
+
+=over 4
+
+=item type
+
+A mandatory event type, which is a simple numeric identity, the
+meaning of which is not known by the event functionality itself.
+
+=item priority
+
+A manatory nonnegative integral quantity.  The lower the priority the earlier
+the event will be processes.
+
```

Comment:
I liked the Richard's suggestion from our discussion that a type might have an inherent priority and we would not need a separate priority value.